t/fits.t

```diff
@@ -2,340 +2,480 @@ use strict;
-    (undef, my $f_out) = File::Temp::tempfile( 'delmeXXXXX', SUFFIX => '.fits', %tmp_opts );
-    my $x = xvals(10);
-    $x->hdr->{'HISTORY'} = $hstr;
-    $x->wfits($f_out);
-    my $xr = rfits($f_out);
-    my $hist = $xr->hdr->{'HISTORY'};
-    $hist = join '', map "$_\n", @$hist if ref $hist eq 'ARRAY';
-    $hist =~ s/ +$//gm;
-    is($hist, $hstr, 'multi-line HISTORY correct with fresh header');
-    (undef, $f_out) = File::Temp::tempfile( 'delmeXXXXX', SUFFIX => '.fits', %tmp_opts ); # new one as Windows unable to remove
+###############################
+subtest 'compressed fits file' => sub {
     my $m51 = rfits('t/m51.fits.fz');
-    $m51->hdr->{HISTORY} = $hstr;
-    $m51->wfits($f_out);
-    my $m51r = rfits($f_out);
-    $hist = $m51r->hdr->{'HISTORY'};
-    $hist = join '', map "$_\n", @$hist if ref $hist eq 'ARRAY';
-    $hist =~ s/ +$//gm;
-    is($hist, $hstr, 'multi-line HISTORY correct with pre-existing header');
-}
+    is_pdl $m51->shape, indx( [ 384, 384 ] ),
+      'right dims from compressed FITS file';
+    my ( $fh, $fname ) = tfile;
+    if ($PDL::Astro_FITS_Header) {
+        my $m51_tbl = rfits( 't/m51.fits.fz', { expand => 0 } );
+        wfits( $m51_tbl, $fname );
+        my $m51_2 = rfits($fname);
+        is_pdl $m51_2, $m51, 'read back written-out bintable FITS file';
+        $m51->wfits( $fname, { compress => 1 } );
+        $m51_2 = rfits($fname);
+        is_pdl $m51_2, $m51, 'read back written-out compressed FITS file';
+        $m51_2->hdrcpy(1);
+        $m51_2                = $m51_2->dummy( 2, 3 )->sever;
+        $m51_2->hdr->{NAXIS}  = 3;
+        $m51_2->hdr->{NAXIS3} = 3;
+        $m51_2->wfits( $fname, { compress => 1 } );
+        my $m51_3 = rfits($fname);
+        is_pdl $m51_3, $m51_2, 'read back written-out compressed RGB FITS file';
+    }
+};
+
+###############################
+subtest 'multi-line HISTORY' => sub {
+
+    my $hstr = join( "\n", 'A' .. 'G', '' );    # must end in newline
+
+    subtest 'fresh header' => sub {
+        my ( $fh, $file ) = tfile;
+        my $x = xvals(10);
+        $x->hdr->{'HISTORY'} = $hstr;
+        $x->wfits($file);
+        my $xr   = rfits($file);
+        my $hist = $xr->hdr->{'HISTORY'};
+        $hist = join '', map "$_\n", @$hist if ref $hist eq 'ARRAY';
+        $hist =~ s/ +$//gm;
+        is( $hist, $hstr, 'correct with fresh header' );
+    };
+
+    subtest 'pre-existing header' => sub {
+        my ( $fh, $file ) = tfile;
+        my $m51 = rfits('t/m51.fits.fz');
+        $m51->hdr->{HISTORY} = $hstr;
+        $m51->wfits($file);
+        my $m51r = rfits($file);
+        my $hist = $m51r->hdr->{'HISTORY'};
+        $hist = join '', map "$_\n", @$hist if ref $hist eq 'ARRAY';
+        $hist =~ s/ +$//gm;
+        is( $hist, $hstr, 'correct with pre-existing header' );
+    };
+
+};
+
+###############################
+subtest 'write null hdu with and without Astro::FITS::Header' => sub {
```

Comment:
The existence of `t/fits-noafh.t` renders this whole section moot. In particular, merely checking that `wfits` didn't throw an exception seems to miss the point a bit. Please remove this change, or make it unconditional on AFH being there, and do an actual check of the read-back data.

Reply:
It does do an actual check of the readback data.  the test for exception is secondary to the check which appears on the next lines.

Reply:
OK. Then please get rid of this whole bit, since exceptions make tests fail anyway.